subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeMatchingConfigurationSelector.java

```diff
@@ -45,13 +44,30 @@
 import java.util.List;
 import java.util.Set;
 
-public abstract class AttributeConfigurationSelector {
+/**
+ * Uses attribute matching to select a list of variants from a graph.
+ *
+ * This class is intentionally named similarly to {@link AttributeMatchingVariantSelector}, as it has a
```

Comment:
Since we're going through the effort to document the naming here, I think its important to mention that this class does a lot more than matching _configurations_. When it was written it probably only matched configurations, but now it is a variant matcher. I think this is important since we're now defining associated exceptions (`AbstractConfigurationSelectionException`) and the term `Configuration` has come up elsewhere in this PR to refer to matching done by this class. 

In general I think dependency-management needs to figure out what a Variant really is. 

In the DSL, we have a `Configuration`, and it can define "variants" based on the `ConfigurationPublications#variant` method. 

In the dependency management engine, we convert the `Configuration`s in the DSL to something a variant (`VariantGraphResolveMetadata`), which itself has sub-variants (`VariantResolveMetadata`)

This naming is incredibly confusing. I don't think this PR will or should resolve this naming conflict, but I see it touching and re-using some legacy naming here so I want to bring it up. 

`Configuration` is a legacy concept. It was borrowed from Ivy, but it is on the way out. In dependency-management, we separate a resolution into _graph resolution_ and _artifact resolution_. `AttributeConfigurationSelector` and `AttributeMatchingVariantSelector` handle variant matching for each phase separately. However it is clear that `AttributeConfigurationSelector` matches variants, based on the type of its methods (it selects among a collection of `VariantGraphResolveState`s)

In terms of something actionable, I would prefer if we didn't proliferate the "Configuration" term here in dependency management. I personally have been referring to these two separate types of selectable things "Variants" and "Artifact Variants", but we also use "Variants" and "Sub Variants". The Artifact Variants / Sub Variants can also be thought of as "Precomputed Transforms", as they are considered in the same place where artifact transforms are applied -- artifact transforms are only applied if we _can't_ find a matching Artifact Variant. 

Sorry for the wall of text. What do you think about revising the naming here?

Reply:
This is a great explanation.  I was aware of this naming mismatch during this work, but didn't want to further complicate this existing refactoring, which is more about changing the structure then comprehensively improving the naming.

I agree `Variants` and `ArtifactVariants` would be much better names and that `Configuration` here is quite confusing.

After this is merged, I'll try renaming these types and methods in a separate PR.

Reply:
"Graph Variants" and "Artifact Variants"?

Reply:
I think this is the way to go. It's better to be explicit here imo